cranelift/interpreter/src/step.rs

```diff
@@ -796,3 +812,78 @@ where
         }
     })
 }
+
+/// Converts a SIMD vector value into a Rust vector of i128 for processing.
+fn extractlanes<V>(x: &V, lane_type: types::Type) -> ValueResult<Vec<i128>>
+where
+    V: Value,
+{
+    let iterations = match lane_type {
+        types::I8 => 1,
+        types::I16 => 2,
+        types::I32 => 4,
+        types::I64 => 8,
+        _ => unimplemented!("Only 128-bit vectors are currently supported."),
+    };
+
+    let x = x.into_array()?;
+    let mut lanes = Vec::new();
+    for (i, _) in x.iter().enumerate() {
+        let mut lane: i128 = 0;
+        if i % iterations != 0 {
+            continue;
+        }
+        for j in 0..iterations {
+            lane += (x[i + j] as i128) << (8 * j);
+        }
+        lanes.push(lane);
+    }
+    return Ok(lanes);
+}
+
+/// Convert a Rust array of i128s back into a `Value::vector`.
+fn vectorizelanes<V>(x: Vec<i128>, vector_type: types::Type) -> ValueResult<V>
+where
+    V: Value,
+{
+    let iterations = match vector_type.lane_type() {
+        types::I8 => 1,
+        types::I16 => 2,
+        types::I32 => 4,
+        types::I64 => 8,
+        _ => unimplemented!("Only 128-bit vectors are currently supported."),
+    };
+    let mut result: [u8; 16] = [0; 16];
+    for (i, val) in x.iter().enumerate() {
+        let val = *val;
+        for j in 0..iterations {
+            result[(i * iterations) + j] = (val >> (8 * j)) as u8;
+        }
+    }
+    Value::vector(result, vector_type)
+}
+
+/// Performs the supplied binary arithmetic `op` on two SIMD vectors.
+fn binary_arith<V, F>(x: V, y: V, vector_type: types::Type, op: F, unsigned: bool) -> ValueResult<V>
```

Comment:
Would it be easier to do a blanket impl of the `Value` trait like: `impl<T: Value, const N: usize> Value for [T; N]`? And have a function that would convert the SIMD values into/from the array?

I don't know how many of these functions we are going to need, but it might be worth considering in the future.

Although I think this would require rust 1.51 with `min_const_generics`. Which might be higher than our minimum requirements.